### PR TITLE
Prevent route loops when provisioning new mappings

### PR DESCRIPTION
**Pull request by lukebarnard1**
The provisioner will fail to request for a mapping if the m.room.bridging status in a room for the same channel (i.e. with the same state key) is not = 'failure'. Meaning, if it is either pending or a success, the request will fail (unless the bridging requester is the same as the ```user_id``` in the ```m.room.bridging```).

Maybe the check for ```user_id``` should really be with the sender, meaning the bridge checks that its own matrix bot was the one to send the bridging. But currently, when you get an event, you don't get it's original sender, so we may have to add extra state to the content of the ```m.room.bridging``` indicating the matrix id of the bridge that set it.

### CODE REVIEW

File: lib/provisioning/Provisioner.js [hunk 1]
```diff
@@ -248,11 +250,46 @@ Provisioner.prototype._authoriseProvisioning = Promise.coroutine(
...
-        // (Matrix) update room state
         // State key for m.room.bridging
         let skey = `irc://${ircDomain}/${ircChannel}`;
+
+        // (Matrix) check room state to prevent route looping
+        try {
+            let bridgingState = yield matrixClient.getStateEvent(roomId, 'm.room.bridging', skey);
+            // Bridging state exists and is either success or pending (ignore failures)
+            if (bridgingState && bridgingState.status !== 'failure') {
+                // If bridging state sender is the bridge matrix user
```
Comment (Kegsay) [inline lib/provisioning/Provisioner.js]: s/bridge matrix user/provisioner/

**Comment by Kegsay** [comment]
LGTM with 1 teeny test :)

**Comment by lukebarnard1** [comment]
@Kegsay as per IRL, PTAL

File: lib/provisioning/Provisioner.js [hunk 2]
```diff
@@ -287,11 +302,84 @@ Provisioner.prototype._authoriseProvisioning = Promise.coroutine(
...
+                (e) => {
+                    return e.type === 'm.room.bridging' && e.state_key === skey
+                }
+            );
+            // Bridging state exists and is either success or pending (ignore failures)
+            if (wholeBridgingState &&
+                wholeBridgingState.content) {
+                let bridgingState = wholeBridgingState.content;
+
+            if (bridgingState.status !== 'failure') {
```
Comment (Kegsay) [inline lib/provisioning/Provisioner.js]: Indentation is weird.

File: lib/provisioning/Provisioner.js [hunk 3]
```diff
@@ -287,11 +302,84 @@ Provisioner.prototype._authoriseProvisioning = Promise.coroutine(
...
+        try {
+            let roomState = yield matrixClient.roomState(roomId);
+            let wholeBridgingState = roomState.find(
+                (e) => {
+                    return e.type === 'm.room.bridging' && e.state_key === skey
+                }
+            );
+            // Bridging state exists and is either success or pending (ignore failures)
+            if (wholeBridgingState &&
+                wholeBridgingState.content) {
```
Comment (Kegsay) [inline lib/provisioning/Provisioner.js]: Double indent please to make it clear where the if body begins.

File: lib/provisioning/Provisioner.js [hunk 4]
```diff
@@ -287,11 +302,84 @@ Provisioner.prototype._authoriseProvisioning = Promise.coroutine(
...
+
+            if (bridgingState.status !== 'failure') {
+                // If bridging state sender is this bot
+                if (wholeBridgingState.sender === matrixClient.credentials.userId) {
+                    // Success, already pending/success
+                    log.info(`Bridging state already exists in room ${roomId} ` +
+                             `(status = ${bridgingState.status},` +
+                             ` bridger = ${bridgingState.user_id}.)`);
+
+                    // If pending, resend the message to the op as if it were the original
```
Comment (Kegsay) [inline lib/provisioning/Provisioner.js]: This comment should be with the if pending block and not the if success block.

File: lib/provisioning/Provisioner.js [hunk 5]
```diff
@@ -287,11 +302,84 @@ Provisioner.prototype._authoriseProvisioning = Promise.coroutine(
...
+                else {// If it is from a different sender, fail
+                    throw new Error(`A request to create this mapping has already been sent ` +
+                             `(status = ${bridgingState.status},` +
+                             ` bridger = ${bridgingState.user_id}. Ignoring request.`);
+                }
+            }
+            }
+        }
+        catch (err) {
+            // The request to discover bridging state has failed
```
Comment (Kegsay) [inline lib/provisioning/Provisioner.js]: Lies. This try block covers more than just getting the bridge state event.

File: lib/provisioning/Provisioner.js [hunk 6]
```diff
@@ -365,20 +461,35 @@ Provisioner.prototype._setRequest = function (server, opNick, request) {
     this._pendingRequests[server.domain][opNick] = request;
 }
 
-Provisioner.prototype.handlePm = function(server, fromUser, text) {
+Provisioner.prototype.handlePm = Promise.coroutine(function*(server, fromUser, text) {
     if (['y', 'yes'].indexOf(text.trim().toLowerCase()) == -1) {
         log.warn(`Provisioner only handles text 'yes'/'y' ` +
                  `(from ${fromUser.nick} on ${server.domain})`);
+
+        this._sendToUser(
```
Comment (Kegsay) [inline lib/provisioning/Provisioner.js]: We should probably yield here so handlePm only returns after we have sent this message.

File: lib/provisioning/Provisioner.js [hunk 7]
```diff
@@ -365,20 +461,35 @@ Provisioner.prototype._setRequest = function (server, opNick, request) {
...
+        yield this._sendToUser(
+            fromUser.nick, server,
+            'Thanks for your response, bridge request authorised.'
+        );
+
         return;
     }
     log.warn(`Provisioner was not expecting PM from ${fromUser.nick} on ${server.domain}`);
-}
+    this._sendToUser(
```
Comment (Kegsay) [inline lib/provisioning/Provisioner.js]: Yield

File: lib/provisioning/Provisioner.js [hunk 8]
```diff
@@ -513,13 +617,41 @@ Provisioner.prototype._doLink = Promise.coroutine(
...
+        if (entry) {
             throw new Error(`Room mapping already exists (${mappingLogId},` +
                             `origin = ${entry.data.origin})`);
         }
+        yield this._ircBridge.getStore().storeRoom(ircRoom, mxRoom, 'provision');
+
+        // Cause the bot to join the new plumbed channel (if it is enabled (see joinChannel))
+        // TODO: key not persisted on restart
+        let botClient = yield this._getBotClientForServer(server);
+        botClient.joinChannel(ircChannel, key);
```
Comment (Kegsay) [inline lib/provisioning/Provisioner.js]: Yield?

File: lib/provisioning/Provisioner.js [hunk 9]
```diff
@@ -551,12 +683,14 @@ Provisioner.prototype.unlink = Promise.coroutine(function*(options) {
...
-    }
-    else {
+    if (!entry) {
         throw new Error(`Provisioned room mapping does not exist (${mappingLogId})`);
     }
+    yield this._ircBridge.getStore().removeRoom(roomId, ircDomain, ircChannel, 'provision');
+
+    // Cause the bot to part the channel (if it is enabled (see leaveChannel))
+    let botClient = yield this._getBotClientForServer(server);
+    botClient.leaveChannel(ircChannel);
```
Comment (Kegsay) [inline lib/provisioning/Provisioner.js]: Yield?

**Comment by Kegsay** [comment]
LGTM aside from comments. Not sure in some places if you should be yielding, but otherwise we're just dropping those promises to the floor without logging errors which is bad.

**Pull request merged**